### PR TITLE
Authenticaton Project Review Request

### PR DESCRIPTION
**Pull request by TashfeenRao**

### CODE REVIEW

File: app/helpers/sessions_helper.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
+      @current ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
```
Comment (samgaco) [inline app/helpers/sessions_helper.rb]: - [ ] We should have to methods, one to retrieve the current user and another one to set it. [Requirement](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png).

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+module SessionsHelper
+  def log_in(user)
+    session[:user_id] = user.id
+  end
```
Comment (samgaco) [inline app/helpers/sessions_helper.rb]: - [ ] We should have a sign-in method. [Requirement](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_method_requirement.png).

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -10,6 +10,9 @@
   </head>
 
   <body>
+    <% flash.each do |message_type, message| %>
+        <%= content_tag(:div, message, class: "alert alert-#{message_type}") %>
+      <% end %>
     <%= yield %>
   </body>
 </html>
```
Comment (samgaco) [inline app/views/layouts/application.html.erb]: - [ ] We should have Sign in/Sign out links present in any view of the app ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_out_link_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49722686/))

File: config/routes.rb [hunk 4]
```diff
@@ -1,3 +1,13 @@
 Rails.application.routes.draw do
+  get 'users/new'
+  get 'user/new'
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  get '/signin', to:'sessions#new'
+  post '/signin', to:'sessions#create'
+  delete '/signout', to:'sessions#destroy'
+  root 'posts#index'
+  get '/posts/new', to: 'posts#new'
```
Comment (samgaco) [inline config/routes.rb]: - [ ] Have this [requirement](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) into account.

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @konyan and @TashfeenRao :high_brightness: 

I had the following error when trying to log into your application (after creating a user on the console):

problemwhenssignin

I also left some comments below that should be addressed.

Good spirit! :muscle: 

-----------------------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+class User < ApplicationRecord
+  attr_accessor :remember_token
+  before_create :remember_token
```
Comment (maelfosso) [inline app/models/user.rb]: Please, use two different names. In your case, I think that what comes close to the `before_create` is something between `digest` and `new_token`

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def self.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def forget
+    update_attribute(:remember_digest, nil)
+  end
```
Comment (maelfosso) [inline app/models/user.rb]: You need to rewrite two 2 methods

`before_create` should only set up `self.remember_token`
You will set up `self.remember_token` with the result of a method which has theses steps
1. Generate a new token
2. Digest the new token and return it.

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
+  end
```
Comment (maelfosso) [inline app/models/user.rb:31]: If you follow the instructions given up, you will update it with only the last lines

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello @TashfeenRao @konyan 

Thank you for the improvements done!
However, there are still things to update

1. Your app lacks navigation. For example, from `Sign In` it's not possible to go back to the list of posts. Another one, once connected, from `posts` there is no link to access to the post creation form. Make sure to simplify the test by implementing a good navigation.

2. There is no user creation form, what is not required and there is no also any indication to how to start using your app. I suggest you update the README with indications on how to start either by creating an user via `rails console` or by using an user you create through `rails db:seed`

3. @samgaco mentionned this https://github.com/TashfeenRao/members-only/pull/1#discussion_r382807836 and you must notice that when implementing it, you no longer need `session`. Try and you will see how `cookies` simplifies your code.

4. Please, read the comments left below.

Kindly make the suggested changes and submit a new code review request.
Thanks

Mael FOSSO

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Thank you for the improvements!!!
I got a problem when launching the app
Screenshot_2020-02-26_00-52-19

Kindly fix that and submit a new code review request.
Thanks

Mael FOSSO

**Review by imhta — APPROVED** [review]
# Project approved 🎉 

### Hello  @TashfeenRao  , 👋 You have done a great job in this project 👏 
- [ ] [ Optional] Pull Request should have a meaningful title and a short description



----
### Happy coding 😄 
----
**Pon Muthu Selvam N**
__nponmuthuselvam@gmail.com__

**Pull request merged**